File: plugins/go/crepuscularity.go
package crepuscularity

import (
	"bytes"
	"encoding/json"
	"fmt"
	"html"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
)

type ViewIr struct {
	Version int              `json:"version"`
	Root    []map[string]any `json:"root"`
}

type Event struct {
	Handler string         `json:"handler"`
	Payload map[string]any `json:"payload,omitempty"`
}

type EventHandler func(Event, *ViewSession) error

type ViewSession struct {
	Path    string
	Context map[string]any
	events  map[string]EventHandler
}

func crepusBin() string {
	if bin := os.Getenv("CREPUS_BIN"); bin != "" {
		if filepath.IsAbs(bin) || filepath.Base(bin) == bin {
			return bin
		}
	}
	return "crepus"
}

var bindBlocklist = map[string]bool{"baseDir": true, "_": true} // ponytail: block security-sensitive keys only

func RenderIR(path string, context map[string]any) (ViewIr, error) {
	cwd, err := os.Getwd()
	if err != nil {
		return ViewIr{}, err
	}

	allowedDir := cwd
	if override := os.Getenv("CREPUS_ALLOWED_DIR"); override != "" {
		allowedDir = override
	}

	absAllowed, err := filepath.Abs(allowedDir)
	if err != nil {
		return ViewIr{}, err
	}

	if evalAllowed, err := filepath.EvalSymlinks(absAllowed); err == nil {
		absAllowed = evalAllowed
	}

	evalPath, err := filepath.EvalSymlinks(path)
	if err != nil {
		if os.IsNotExist(err) {
			evalPath = path
		} else {
			return ViewIr{}, err
		}
	}

	absPath, err := filepath.Abs(evalPath)
	if err != nil {
		return ViewIr{}, err
	}

	rel, err := filepath.Rel(absAllowed, absPath)
	if err != nil {
		return ViewIr{}, err
	}
	if rel == ".." || strings.HasPrefix(rel, "../") || strings.HasPrefix(rel, "..\\") {
		return ViewIr{}, fmt.Errorf("path traversal denied")
	}

	payload, err := json.Marshal(map[string]any{
		"template": mustRead(absPath),
		"context":  context,
	})
	if err != nil {
		return ViewIr{}, err
	}
	cmd := exec.Command(crepusBin(), "native", "ir", "--stdin-json")
	cmd.Stdin = bytes.NewReader(payload)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return ViewIr{}, fmt.Errorf("%w: %s", err, string(out))
	}
	var ir ViewIr
	if err := json.Unmarshal(out, &ir); err != nil {
		return ViewIr{}, err
	}
	return ir, nil
}

func mustRead(path string) string {
	data, err := os.ReadFile(path)
	if err != nil {
		panic(err)
	}
	return string(data)
}

func RenderHTML(path string, context map[string]any) (string, error) {
	ir, err := RenderIR(path, context)
	if err != nil {
		return "", err
	}
	var out strings.Builder
	for _, node := range ir.Root {
		out.WriteString(renderNode(node))
	}
	return out.String(), nil
}

func NewViewSession(path string, context map[string]any) *ViewSession {
	if context == nil {
		context = map[string]any{}
	}
	return &ViewSession{Path: path, Context: context, events: map[string]EventHandler{}}
}

func (session *ViewSession) On(handler string, callback EventHandler) *ViewSession {
	session.events[handler] = callback
	return session
}

func (session *ViewSession) RenderIR() (ViewIr, error) {
	return RenderIR(session.Path, session.Context)
}

func (session *ViewSession) RenderHTML() (string, error) {
	return RenderHTML(session.Path, session.Context)
}

func (session *ViewSession) Dispatch(event Event) (ViewIr, error) {
	if strings.HasPrefix(event.Handler, "bind:") {
		parts := strings.SplitN(strings.TrimPrefix(event.Handler, "bind:"), ":", 2)
		// ponytail: only allowlisted keys from bind:
		if len(parts) == 2 && !bindBlocklist[parts[0]] {
			session.Context[parts[0]] = parts[1]
		}
	}
	if callback := session.events[event.Handler]; callback != nil {
		if err := callback(event, session); err != nil {
			return ViewIr{}, err
		}
	}
	return session.RenderIR()
}

func renderNode(node map[string]any) string {
	switch node["kind"] {
	case "text":
		return html.EscapeString(fmt.Sprint(node["content"]))
	case "stack", "scroll":
		var out strings.Builder
		out.WriteString(`<div data-crepus-kind="`)
		out.WriteString(html.EscapeString(fmt.Sprint(node["kind"])))
		out.WriteString(`" data-axis="`)
		out.WriteString(html.EscapeString(fmt.Sprint(node["axis"])))
		out.WriteString(`">`)
		for _, child := range nodeList(node["children"]) {
			out.WriteString(renderNode(child))
		}
		out.WriteString("</div>")
		return out.String()
	case "button":
		label := html.EscapeString(fmt.Sprint(node["label"]))
		if node["onClick"] != nil {
			return `<button data-onclick="` + html.EscapeString(fmt.Sprint(node["onClick"])) + `">` + label + `</button>`
		}
		return "<button>" + label + "</button>"
	case "image":
		return `<img src="` + html.EscapeString(fmt.Sprint(node["src"])) + `" alt="` + html.EscapeString(fmt.Sprint(node["alt"])) + `">`
	case "slotRotate":
		phrases := valueList(node["phrases"])
		if len(phrases) == 0 {
			return `<span data-crepus-kind="slotRotate"></span>`
		}
		return `<span data-crepus-kind="slotRotate">` + html.EscapeString(fmt.Sprint(phrases[0])) + `</span>`
	case "input":
		bind := html.EscapeString(fmt.Sprint(node["bind"]))
		placeholder := html.EscapeString(fmt.Sprint(node["placeholder"]))
		if node["multiline"] == true {
			return `<textarea data-bind="` + bind + `" placeholder="` + placeholder + `"></textarea>`
		}
		return `<input data-bind="` + bind + `" placeholder="` + placeholder + `">`
	case "picker":
		var out strings.Builder
		out.WriteString(`<select data-bind="`)
		out.WriteString(html.EscapeString(fmt.Sprint(node["bind"])))
		out.WriteString(`">`)
		for _, option := range nodeList(node["options"]) {
			out.WriteString(`<option value="`)
			out.WriteString(html.EscapeString(fmt.Sprint(option["value"])))
			out.WriteString(`">`)
			out.WriteString(html.EscapeString(fmt.Sprint(option["label"])))
			out.WriteString(`</option>`)
		}
		out.WriteString(`</select>`)
		return out.String()
	default:
		return ""
	}
}

func nodeList(value any) []map[string]any {
	raw, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]map[string]any, 0, len(raw))
	for _, item := range raw {
		if node, ok := item.(map[string]any); ok {
			out = append(out, node)
		}
	}
	return out
}

func valueList(value any) []any {
	raw, _ := value.([]any)
	return raw
}

File: plugins/zig/crepuscularity.zig
const std = @import("std");

pub const ViewIr = struct {
    version: u32,
    json: []u8,
};

pub const UiDocument = struct {
    html: []const u8,
};

pub fn renderIr(allocator: std.mem.Allocator, path: []const u8) !ViewIr {
    // ponytail: argv exec, no shell
    const result = try std.process.Child.run(.{
        .allocator = allocator,
        .argv = &.{ "crepus", "native", "ir", path },
    });
    defer allocator.free(result.stderr);
    if (result.term.Exited != 0) return error.CrepusFailed;
    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, result.stdout, .{});
    defer parsed.deinit();
    const version = @as(u32, @intCast(parsed.value.object.get("version").?.integer));
    return .{ .version = version, .json = result.stdout };
}

pub fn renderHtml(allocator: std.mem.Allocator, path: []const u8) !UiDocument {
    var ir = try renderIr(allocator, path);
    defer allocator.free(ir.json);

    const parsed = try std.json.parseFromSlice(std.json.Value, allocator, ir.json, .{});
    defer parsed.deinit();

    const root = parsed.value.object.get("root") orelse return error.MissingRoot;
    var html_buf = std.ArrayList(u8).init(allocator);
    defer html_buf.deinit();
    try renderNodeToHtml(root.*, &html_buf);

    return .{ .html = try html_buf.toOwnedSlice(allocator) };
}

fn renderNodeToHtml(node: std.json.Value, buf: *std.ArrayList(u8)) !void {
    if (node != .object) return;
    const kind = if (node.object.get("kind")) |k| switch (k) { .string => |s| s, else => return } else return;

    if (std.mem.eql(u8, kind, "text")) {
        const content = if (node.object.get("content")) |c| switch (c) { .string => |s| s, else => "" } else "";
        try buf.appendSlice(content);
    } else if (std.mem.eql(u8, kind, "stack")) {
        const axis = if (node.object.get("axis")) |a| switch (a) { .string => |s| s, else => "column" } else "column";
        try buf.writer().print("<div data-crepus-kind=\"stack\" data-axis=\"{s}\">", .{axis});
        try renderChildren(node, buf);
        try buf.appendSlice("</div>");
    } else if (std.mem.eql(u8, kind, "button")) {
        const label = if (node.object.get("label")) |l| switch (l) { .string => |s| s, else => "" } else "";
        try buf.writer().print("<button>{s}</button>", .{label});
    } else if (std.mem.eql(u8, kind, "scroll")) {
        const axis = if (node.object.get("axis")) |a| switch (a) { .string => |s| s, else => "column" } else "column";
        try buf.writer().print("<div data-crepus-kind=\"scroll\" data-axis=\"{s}\">", .{axis});
        try renderChildren(node, buf);
        try buf.appendSlice("</div>");
    } else if (std.mem.eql(u8, kind, "image")) {
        const src = if (node.object.get("src")) |s| switch (s) { .string => |v| v, else => "" } else "";
        try buf.writer().print("<img src=\"{s}\">", .{src});
    }
}

fn renderChildren(node: std.json.Value, buf: *std.ArrayList(u8)) !void {
    const children = if (node.object.get("children")) |c| switch (c) {
        .array => |a| a.items,
        else => return,
    } else return;
    for (children) |child| {
        try renderNodeToHtml(child, buf);
    }
}

test "renderIr decodes ViewIr" {
    const allocator = std.testing.allocator;
    const ir = try renderIr(allocator, "../fixtures/hello.crepus");
    defer allocator.free(ir.json);
    try std.testing.expectEqual(@as(u32, 6), ir.version);
}

test "renderHtml outputs valid HTML" {
    const allocator = std.testing.allocator;
    const doc = try renderHtml(allocator, "../fixtures/hello.crepus");
    defer allocator.free(doc.html);
    try std.testing.expect(doc.html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, doc.html, "Hello") != null);
}

test "renderHtml handles interactive template" {
    const allocator = std.testing.allocator;
    const doc = try renderHtml(allocator, "../fixtures/interactive.crepus");
    defer allocator.free(doc.html);
    try std.testing.expect(doc.html.len > 0);
    try std.testing.expect(std.mem.indexOf(u8, doc.html, "button") != null);
}
